android/app/src/main/java/com/aws/greengrass/nucleus/androidservice/MainActivity.java

```diff
@@ -0,0 +1,325 @@
+
+public class MainActivity extends AppCompatActivity {
+
+    private ActivityMainBinding binding;
+
+    private final ExecutorService backgroundExecutor = Executors.newSingleThreadExecutor();
+    private ProvisionManager provisionManager;
+    private ServicesConfigurationProvider servicesConfigProvider;
+    private Executor mainExecutor = null;
+    private JsonNode provisioningConfig = null;
+    private static Logger logger;
+
+    private final String thingNameStr = "Thing name:\n";
+
+    private final ActivityResultLauncher<Intent> provisioningConfigResultLauncher = registerForActivityResult(
+            new ActivityResultContracts.StartActivityForResult(),
+            result -> {
+                if (result.getResultCode() == Activity.RESULT_OK) {
+                    Intent data = result.getData();
+                    if (data != null) {
+                        processFile(data.getData());
+                    }
+                }
+            });
+
+    private final ActivityResultLauncher<Intent> servicesConfigResultLauncher = registerForActivityResult(
+            new ActivityResultContracts.StartActivityForResult(),
+            result -> {
+                if (result.getResultCode() == Activity.RESULT_OK) {
+                    Intent data = result.getData();
+                    if (data != null) {
+                        try {
+                            servicesConfigProvider.setExternalConfig(getContentResolver()
+                                    .openInputStream(data.getData()));
+                        } catch (FileNotFoundException e) {
+                            logger.atError().setCause(e).log("File isn't found.");
+                        }
+                    }
+                }
+            });
+
+    @Override
+    protected void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        if (logger == null) {
+            logger = LogHelper.getLogger(this.getFilesDir(), MainActivity.class);
+        }
+        provisionManager = BaseProvisionManager.getInstance(getFilesDir());
+        servicesConfigProvider = ServicesConfigurationProvider.getInstance(getFilesDir());
+        mainExecutor = ContextCompat.getMainExecutor(this);
+        binding = ActivityMainBinding.inflate(getLayoutInflater());
+        View view = binding.getRoot();
+        setContentView(view);
+
+        bindConfigUI();
+        bindStartStopUI();
+        binding.version.setText(getString(R.string.label_app_version, BuildConfig.VERSION_NAME));
+
+        if (provisionManager.isProvisioned()) {
+            binding.thingNameText.setText(String.format("%s%s",
+                    thingNameStr, provisionManager.getThingName()));
+            switchUI(false);
+        } else {
+            provisionManager.clearProvision();
+            switchUI(true);
+        }
+    }
+
+    @Override
+    protected void onDestroy() {
+        backgroundExecutor.shutdownNow();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Call `shutdown()` to reject any incoming tasks and `awaitTermination()` to wait for existing tasks to complete or until a timeout is reached. To stop the execution and processing of tasks, if necessary, call `shutdownNow()`. Check the return value of `awaitTermination()` and handle when the method times out while waiting for other threads to stop executing. See the following diffs on GitHub that make this code change: [diff1](https://github.com/glowroot/glowroot/commit/81de5a3699c59e0cf98d485c380c1034e6cd28a5#diff-c608d2c40790b0e0e5228e6647f69729caa756bbe9127169139406314b80897eL143), [diff2](https://github.com/apache/directory-kerby/commit/3ed0e7cd3b8a0c3ce29293153e4e72f422f0637f#diff-eb42f072849f520563ec2a686789a836195f18075cc222ef355c6ea710b0a360L84)

Learn more: [ExecutorService in the Java Platform, Standard Edition 8 API Specification](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ExecutorService.html)